IPE-1607: Updating Consul workload install: clients and dataplane (#3132)
hashicorp/web-unified-docs: content/consul/v2.0.x/content/docs/deploy/workload/client/vm.mdx

- CA certificate
- TLS server name
- Valid ACL token

For more information, refer to the [Consul agents configuration file reference](/consul/docs/reference/agent/configuration-file).

## Joining a Consul cluster on K8s

Consul agents running on VMs can also join a Consul server cluster running within Kubernetes.

### Auto-join

> The recommended way to join a cluster running within Kubernetes is to use the ["k8s" cloud auto-join provider](/consul/docs/deploy/server/cloud-auto-join#kubernetes-k8s).

Reviewer: ```suggestion
We recommend using the ["k8s" cloud auto-join provider](/consul/docs/deploy/server/cloud-auto-join#kubernetes-k8s) to join a cluster running within Kubernetes.
```

Style guide:  active voice, point of view

> The auto-join provider dynamically discovers IP addresses to join using the Kubernetes API. It authenticates with Kubernetes using a standard `kubeconfig` file. Auto-join works with all major hosted Kubernetes offerings as well as self-hosted installations. The token in the `kubeconfig` file needs to have permissions to list pods in the namespace where Consul servers

Reviewer: ```suggestion
The auto-join provider dynamically discovers IP addresses to join using the Kubernetes API. It authenticates with Kubernetes using a standard `kubeconfig` file. Auto-join works with all major hosted Kubernetes offerings as well as self-hosted installations. The token in the `kubeconfig` file must have permissions to list pods in the namespace where Consul servers
```

are deployed.

> The auto-join string below joins a Consul server agent to a cluster using the [official Helm chart](/consul/docs/reference/k8s/helm):

Reviewer: ```suggestion
The following auto-join command joins a Consul server agent to a cluster using the [official Helm chart](/consul/docs/reference/k8s/helm):
```

```shell-session
$ consul agent -retry-join 'provider=k8s label_selector="app=consul,component=server"'
```

> -> **Note:** This auto-join command only connects on the default gossip port 8301, whether you are joining on the pod network or via host ports. A Consul server that is already a member of the datacenter should be listening on this port for the external service to connect through auto-join.

Reviewer: I think the intent of this note is to give two requirements:
1. This uses the default port
2. You need a Consul server already running

We could rewrite this into one directive instruction, something like this, probably put just ahead of the command.

> Before using the auto-join command, ensure that a Consul server is listening on the default gossip port of 8301.

(I am not fluent enough in Consul to know whether we have to say something like "a Consul server in the datacenter")

Author: Good point! I will implement this

### Auto-join on the Pod network

> In the default Consul Helm chart installation, Consul servers are routable through their pod IPs for server RPCs. As a result, any external agents joining the Consul cluster running on Kubernetes need to be able to connect to those pod IPs.

Reviewer: ```suggestion
The default Consul Helm chart installation makes Consul servers  routable through their pod IPs for server RPCs. As a result, any external agents joining the Consul cluster running on Kubernetes need to be able to connect to those pod IPs.
```

Switching to active voice

In many hosted Kubernetes environments, you need to explicitly configure your hosting provider to ensure that pod IPs are routable from external VMs.

> For more information, refer to [Azure AKS CNI](https://docs.microsoft.com/en-us/azure/aks/concepts-network#azure-cni-advanced-networking), [AWS EKS CNI](https://docs.aws.amazon.com/eks/latest/userguide/pod-networking.html) and [GKE VPC-native clusters](https://cloud.google.com/kubernetes-engine/docs/concepts/alias-ips).

Reviewer: ```suggestion
For more information, refer to hosting provider documentation, such as [Azure AKS CNI](https://docs.microsoft.com/en-us/azure/aks/concepts-network#azure-cni-advanced-networking), [AWS EKS CNI](https://docs.aws.amazon.com/eks/latest/userguide/pod-networking.html) and [GKE VPC-native clusters](https://cloud.google.com/kubernetes-engine/docs/concepts/alias-ips).
```

  1. On the external VM, run:

  ```shell-session
  consul agent \
    -advertise="$ADVERTISE_IP" \
    -retry-join='provider=k8s label_selector="app=consul,component=server"' \
    -bind=0.0.0.0 \
    -hcl='leave_on_terminate = true' \
    -hcl='ports { grpc = 8502 }' \
    -config-dir=$CONFIG_DIR \
    -datacenter=$DATACENTER \
    -data-dir=$DATA_DIR \
  ```

  1. Run `consul members` to check if the join was successful.

  ```shell-session
  / $ consul members
  Node                                           Address           Status  Type    Build  Protocol  DC   Segment
  consul-consul-server-0                         10.138.0.43:9301  alive    server  1.9.1  2         dc1  <all>
  external-agent                                 10.138.0.38:8301  alive    client  1.9.0  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-grs4  10.138.0.43:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-otge  10.138.0.44:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-vo7k  10.138.0.42:8301  alive    client  1.9.1  2         dc1  <default>
  ```

### Auto-join through host ports

> If your external VMs cannot connect to Kubernetes pod IPs but they can connect to the internal host IPs of the nodes in the Kubernetes cluster, you can join the two by exposing ports on the host IP instead.

Reviewer: "join the two"...two whats?  Clusters?  Agents?  Agent to the cluster?

This is me not knowing things, but for people that don't know things, it's helpful to have it spelled out. :)

    -retry-join='provider=k8s host_network=true label_selector="app=consul,component=server"'
    -bind=0.0.0.0 \
    -hcl='leave_on_terminate = true' \
    -hcl='ports { grpc = 8502 }' \
    -config-dir=$CONFIG_DIR \
    -datacenter=$DATACENTER \
    -data-dir=$DATA_DIR \
  ```

  Note the addition of `host_network=true` in the retry-join argument.

  1. Run `consul members` to check if the join was successful.
   
  ```shell-session
  / $ consul members
  Node                                           Address           Status  Type    Build  Protocol  DC   Segment
  consul-consul-server-0                         10.138.0.43:9301  alive    server  1.9.1  2         dc1  <all>
  external-agent                                 10.138.0.38:8301  alive    client  1.9.0  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-grs4  10.138.0.43:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-otge  10.138.0.44:8301  alive    client  1.9.1  2         dc1  <default>
  gke-external-agent-default-pool-32d15192-vo7k  10.138.0.42:8301  alive    client  1.9.1  2         dc1  <default>
  ```

### Manual join

> If you are unable to use auto-join, try following the instructions in either of the auto-join sections, but instead of using a `provider` key in the `-retry-join` flag, pass the address of at least one Consul server. Example: `-retry-join=$CONSUL_SERVER_IP:$SERVER_SERFLAN_PORT`.

Reviewer: ```suggestion
If you are unable to use auto-join, you can manually join your VM to the Consul cluster. Follow the instructions in either of the auto-join sections, but instead of using a `provider` key in the `-retry-join` flag, pass the address of at least one Consul server. Example: `-retry-join=$CONSUL_SERVER_IP:$SERVER_SERFLAN_PORT`.
```

Would it be worth giving the full command as well?

A `kubeconfig` file is not required when using manual join.

> Instead of hardcoding an IP address, we recommend you set up a DNS entry that resolves to the pod IPs or host IPs that the Consul server pods are running on.

Reviewer: If our recommendation is to use DNS instead of IPs, should the example above read `-retry-join=$CONSUL_SERVER_NAME:$SERVER_SERFLAN_PORT`?

(Or `CONSUL_SERVER_ADDRESS`, which leaves it at least ambiguous?)
